lib/model/model.go

```diff
@@ -1777,32 +1879,54 @@ func (m *model) introduceDevice(device protocol.Device, introducerCfg config.Dev
 
 // Closed is called when a connection has been closed
 func (m *model) Closed(conn protocol.Connection, err error) {
-	device := conn.DeviceID()
+	connID := conn.ConnectionID()
+	deviceID := conn.DeviceID()
+
 	m.pmut.Lock()
-	conn, ok := m.conn[device]
+	conn, ok := m.connections[connID]
 	if !ok {
 		m.pmut.Unlock()
 		return
 	}
 
-	delete(m.conn, device)
-	delete(m.connRequestLimiters, device)
-	delete(m.helloMessages, device)
-	delete(m.deviceDownloads, device)
-	delete(m.remoteFolderStates, device)
-	closed := m.closed[device]
-	delete(m.closed, device)
-	delete(m.indexHandlers, device)
+	closed := m.closed[connID]
+	delete(m.closed, connID)
+	delete(m.connections, connID)
+
+	removedIsPrimary := m.promotedConnID[deviceID] == connID
+	remainingConns := without(m.deviceConnIDs[deviceID], connID)
+	if removedIsPrimary {
+		m.progressEmitter.temporaryIndexUnsubscribe(conn)
+		if idxh, ok := m.indexHandlers[deviceID]; ok && idxh.conn.ConnectionID() == connID {
+			delete(m.indexHandlers, deviceID)
+		}
```

Comment:
This might be for the case where we've moved the index handler to a different, supposedly non-promoted, connection due to getting index updates there instead... If so, it couldn't hurt with some commentary.

Yes, I was also a little bit confused about the index handler lifetime.

Reply:
> This might be for the case where we've moved the index handler to a different, supposedly non-promoted, connection due to getting index updates there instead... If so, it couldn't hurt with some commentary.

Definitely need to look into that (moving index handler), I managed to not even notice that behaviour/path while reviewing until you mentioned it in a comment above and here.

Reply:
Oh also the index handler stuff is the (or at last a) cause for the flaky cluster config test that came up somewhere else - I'll PR a fix for that, the conflicts won't be bad.

Reply:
I think this is fine now. One thing I may want to do in the long run, but which didn't look trivial right now, is to separate the sending of indexes with the receiving. These are in principle two separate processes that don't need to be intertwined, but currently they are both bundled in the indexhandler. If they were separated then it wouldn't matter at all if both sides agreed on a "primary" connection, other than for the slight latency advantage it might have to avoid blocking the index-sending connection with data-sending.